test/support/iterator_utils.h

```diff
@@ -42,7 +42,19 @@ class ForwardIterator
     ForwardIterator() = default;
     explicit ForwardIterator(Iterator i) : my_iterator(i) {}
     reference operator*() const { return *my_iterator; }
-    Iterator operator->() const { return my_iterator; }
+    pointer
```

Comment:
or like this:
```
    pointer
    operator->() const
    {
        return ::std::addressof(this->operator*());
    }
```

Reply:
Yes, this is better and I believe it works.  I'll check a few things then make the change.  Thanks.

Reply:
Well, I think the specialization for pointers was really the right approach. ``std::adderssof`` always returns a raw pointer, but what if ``operator->`` returns something more complicated, such as a pair of {base pointer, offset} or a smart pointer?

Reply:
Unfortunately I think we made more then required - please see partially reverted changes in https://github.com/oneapi-src/oneDPL/pull/767

Reply:
I believe the intention of this class is to restrict an input iterator to to LegacyForwardIterator, and according to the semantics of [LegacyInputIterator,](https://en.cppreference.com/w/cpp/named_req/InputIterator) and as I understand anything can be returned from `operator->` as long as you can then recurse and call `operator->` on the return until you end up at:

`i->m`  being equivalent to `(*i).m`

If `pointer` is defined as something other than a raw pointer, then this could get us into trouble (as the return would not match the `pointer` type, and the chaining and specialization for pointers was the right way to go.  
I don't think we use it for any exotic pointer types like that in our test suite now, but nothing in the code prevents it.

I can open a new PR to return to the previous approach.